sql/catalyst/src/main/scala/org/apache/spark/sql/types/StructType.scala

```diff
@@ -559,6 +559,42 @@ object StructType extends AbstractDataType {
       case _ => dt
     }
 
+  /**
+   * This works a little similarly to `merge`, but it does not actually merge two DataTypes.
+   * This method just merges nullability.
+   */
+  private[sql] def mergeNullability(left: DataType, right: DataType): DataType =
+    (left, right) match {
+      case (ArrayType(leftElementType, leftContainsNull),
+          ArrayType(rightElementType, rightContainsNull)) =>
+        ArrayType(
+          merge(leftElementType, rightElementType),
+          leftContainsNull || rightContainsNull)
+
+      case (MapType(leftKeyType, leftValueType, leftContainsNull),
+          MapType(rightKeyType, rightValueType, rightContainsNull)) =>
+        MapType(
+          merge(leftKeyType, rightKeyType),
+          merge(leftValueType, rightValueType),
+          leftContainsNull || rightContainsNull)
+
+      case (StructType(leftFields), StructType(rightFields)) =>
+        require(leftFields.size == rightFields.size, "To merge nullability," +
+          "two structs must have same number of fields.")
+
+        val newFields = leftFields.zip(rightFields).map {
```

Comment:
This looks like a regression because this code assume that two structs have identical orders. The previous code doesn't. Could you compare with `merge` and add a test case for the different ordering?

Reply:
For two structs with different order (same or not same types), `CheckAnalysis` will fail such union query. E.g.,

```
[info]   org.apache.spark.sql.AnalysisException: Union can only be performed on tables with the compatible column types. struct<inner2:bigint,col2:struct<inner1:bigint>> <> struct<inner1:bigi
nt,col2:struct<inner2:bigint>> at the second column of the second table;              
```
or

```
[info]   org.apache.spark.sql.AnalysisException: Union can only be performed on tables with the compatible column types. struct<col1:string,col2:struct<inner1:bigint>> <> struct<inner1:bigint
,col2:struct<col1:string>> at the second column of the second table;                                
```

Reply:
Let me add test case for it.